core/src/tx.rs

```diff
@@ -127,8 +118,30 @@ impl TransactionValidator {
             }));
         }
 
-        self.validate_with_builtin_validators(&tx, wsv, is_genesis)?;
-        Self::validate_with_runtime_validators(tx, wsv)
+        if !is_genesis {
+            debug!("Validating transaction: {:?}", tx);
+
+            let AcceptedTransaction {
+                payload,
+                signatures,
+            } = tx.clone();
+            let signatures = signatures.into_iter().collect();
+
+            let signed_tx = SignedTransaction {
+                payload,
+                signatures,
+            };
+            Self::validate_with_runtime_validators(
+                signed_tx.payload.account_id.clone(),
+                signed_tx,
+                wsv,
+            )?;
+        }
+
+        self.validate_and_execute(tx, wsv, is_genesis)?;
```

Comment:
are we validating twice here for non genesis transactions?

Reply:
Indeed it's confusing. I changed the name to `validate_and_execute_instructions()`.

So the current design expects that we can have a lot of validators kinds. Some of them check transactions, some of them check instractions and etc.
On lines 121-139 we check the whole transaction (although we don't have such validators for now). And on line 141 we check every instruction inside transaction.

This idea should be redesigned according to the new idea of one validator. I think, this could be done in #3231